src/Input/Mouse.php

```diff
@@ -181,13 +193,64 @@ private function scroll(int $distance)
             'type' => 'mouseWheel',
             'x' => $this->x,
             'y' => $this->y,
-            'deltaX' => 0,
-            'deltaY' => $distance,
+            'deltaX' => $distanceX,
+            'deltaY' => $distanceY,
         ]));
 
+        // wait until the scroll is done
+        Utils::tryWithTimeout(30000 * 1000, $this->waitForScroll($targetX, $targetY));
+
         // set new position after move
-        $this->y += $distance;
+        $this->x += $distanceX;
+        $this->y += $distanceY;
 
         return $this;
     }
+
+    /**
+     * Get the maximum distance to scroll a page.
+     *
+     * @param int $distance Distance to scroll, positive or negative
+     * @param int $current  Current position
+     * @param int $maximum  Maximum posible distance
+     *
+     * @return int allowed distance to scroll
+     */
+    private function getMaximumDistance(int $distance, int $current, int $maximum): int
+    {
+        $result = $current + $distance;
+
+        if ($result < 0) {
+            return $distance + \abs($result);
+        }
+
+        if ($result > $maximum) {
+            return $maximum - $current;
+        }
+
+        return $distance;
+    }
+
+    /**
+     * Wait for the browser to process the scroll command.
+     *
+     * @param int $targetX
+     * @param int $targetY
+     *
+     * @throws \HeadlessChromium\Exception\OperationTimedOut
+     *
+     * @return bool|\Generator
+     */
+    private function waitForScroll(int $targetX, int $targetY)
+    {
+        while (true) {
+            $visibleArea = $this->page->getLayoutMetrics()->getVisualViewport();
+
+            if ($visibleArea['pageX'] === $targetX && $visibleArea['pageY'] === $targetY) {
+                return true;
+            }
+
+            yield 1000;
```

Comment:
what does this do?

Reply:
(the `1000`, not the `yield` :trollface:)

Reply:
It's the number of microseconds to wait before retry, for the `\HeadlessChromium\Utils::tryWithTimeout` method. I'll update the docblock.